// @api-platform/admin/lib/layout/AppBar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactAdmin = require("react-admin");

var _core = require("@material-ui/core");

var _Logo = _interopRequireDefault(require("./Logo"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

const styles = {
  title: {
    flex: 1,
    textOverflow: 'ellipsis',
    whiteSpace: 'nowrap',
    overflow: 'hidden'
  },
  spacer: {
    flex: 1
  }
};
const CustomAppBar = (0, _core.withStyles)(styles)(({
  classes,
  userMenu,
  ...props
}) => {
  const authProvider = (0, _reactAdmin.useAuthProvider)();
  return /*#__PURE__*/_react.default.createElement(_reactAdmin.AppBar, _extends({
    userMenu: userMenu || !!authProvider
  }, props), /*#__PURE__*/_react.default.createElement(_core.Typography, {
    variant: "h6",
    color: "inherit",
    className: classes.title,
    id: "react-admin-title"
  }), /*#__PURE__*/_react.default.createElement(_Logo.default, null), /*#__PURE__*/_react.default.createElement("span", {
    className: classes.spacer
  }));
});
var _default = CustomAppBar;
exports.default = _default;
